src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,197 @@
+   * @param options.rateLimitTimeout - The time window in which the rate limit is applied.
+   * @param options.rateLimitCount - The amount of notifications an origin can show in the rate limit time window.
+   */
+  constructor({
+    // @todo Pick some sane defaults for this
+    rateLimitTimeout = 5000,
+    rateLimitCount = 3,
+    messenger,
+    state,
+    implementations,
+  }: {
+    rateLimitTimeout?: number;
+    rateLimitCount?: number;
+    messenger: RateLimitMessenger<RateLimitedApis>;
+    state?: Partial<RateLimitState<RateLimitedApis>>;
+    implementations: RateLimitedApis;
+  }) {
+    const defaultState = {
+      requests: Object.keys(implementations).reduce(
+        (acc, key) => ({ ...acc, [key]: {} }),
+        {} as Record<keyof RateLimitedApis, Record<string, number>>,
+      ),
+    };
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.implementations = implementations;
+    this.rateLimitTimeout = rateLimitTimeout;
+    this.rateLimitCount = rateLimitCount;
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:call` as const,
+      (
+        origin: string,
+        type: keyof RateLimitedApis,
+        ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]>
+      ) => this.call(origin, type, ...args),
+    );
+  }
+
+  /**
+   * Calls an API if the requesting origin is not rate-limited.
+   *
+   * @param origin - The requesting origin.
+   * @param type - The type of API call to make.
+   * @param args - Arguments for the API call.
+   * @returns `false` if rate-limited, and `true` otherwise.
+   */
+  async call(
+    origin: string,
+    type: keyof RateLimitedApis,
+    ...args: Parameters<RateLimitedApis[keyof RateLimitedApis]>
+  ): Promise<RateLimitWrapper> {
+    if (this._isRateLimited(type, origin)) {
+      return { isRateLimited: true };
+    }
+    this._recordRequest(type, origin);
+
+    const implementation = this.implementations[type];
+
+    if (!implementation) {
+      throw new Error('Invalid api type');
+    }
+
+    const result = await implementation(...args);
+
+    return { isRateLimited: false, result };
+  }
+
+  /**
+   * Checks whether an origin is rate limited for the a specific API.
+   *
+   * @param api - The API the origin is trying to access.
+   * @param origin - The origin trying to access the API.
+   * @returns `true` if rate-limited, and `false` otherwise.
+   */
+  _isRateLimited(api: keyof RateLimitedApis, origin: string) {
+    return this.state.requests[api][origin] >= this.rateLimitCount;
+  }
+
+  /**
+   * Records that an origin has made a request to call an API, for rate-limiting purposes.
+   *
+   * @param api - The API the origin is trying to access.
+   * @param origin - The origin trying to access the API.
+   */
+  _recordRequest(api: keyof RateLimitedApis, origin: string) {
+    this.update((state) => {
+      (state as any).requests[api][origin] =
```

Comment:
Why do we have to use `as any` here?

Reply:
I recommended this because `immer`'s `Draft` and `WritableDraft` types have issues with generics. @Gudahtt, @shanejonas, and I have fought with this many times. I would welcome any help to get us to a better place with this!

Reply:
Yep, otherwise completely valid types seem to throw errors because of Immer

Reply:
I tried to dig into this and was not able to get this to work without the cast, nor was I able to understand why it failed. I don't think the draft is the issue here, it's something about the index types and `Record`. I type casted the Draft type away and the problem remained.